lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java

```diff
@@ -178,42 +192,127 @@ public ColumnChunk readPrimitive(Field field)
             definitionLevel = definitionReader.readLevel();
             repetitionLevel = repetitionReader.readLevel();
         }
+        int rowCount = 0;
         int valueCount = 0;
         for (int i = 0; i < valuesToRead; i++) {
             do {
+                nextRow(repetitionLevel);
+                if (repetitionLevel == 0) {
+                    rowCount++;
+                }
                 valueReader.run();
                 valueCount++;
                 if (valueCount == remainingValueCountInPage) {
-                    updateValueCounts(valueCount);
+                    updateValueCounts(valueCount, 0);
+                    if (!readNextPage()) {
+                        return rowCount;
+                    }
+                    valueCount = 0;
+                }
+                repetitionLevel = repetitionReader.readLevel();
+                definitionLevel = definitionReader.readLevel();
+            }
+            while (repetitionLevel != 0);
+        }
+        updateValueCounts(valueCount, 0);
+        return rowCount;
+    }
+
+    /**
+     * When filtering using column indexes we might skip reading some pages for different columns. Because the rows are
+     * not aligned between the pages of the different columns it might be required to skip some values. The values (and the
+     * related rl and dl) are skipped based on the iterator of the required row indexes and the first row index of each
+     * page.
+     * For example:
+     *
+     * rows   col1   col2   col3
+     *      ┌──────┬──────┬──────┐
+     *   0  │  p0  │      │      │
+     *      ╞══════╡  p0  │  p0  │
+     *  20  │ p1(X)│------│------│
+     *      ╞══════╪══════╡      │
+     *  40  │ p2(X)│      │------│
+     *      ╞══════╡ p1(X)╞══════╡
+     *  60  │ p3(X)│      │------│
+     *      ╞══════╪══════╡      │
+     *  80  │  p4  │      │  p1  │
+     *      ╞══════╡  p2  │      │
+     * 100  │  p5  │      │      │
+     *      └──────┴──────┴──────┘
+     *
+     * The pages 1, 2, 3 in col1 are skipped so we have to skip the rows [20, 79]. Because page 1 in col2 contains values
+     * only for the rows [40, 79] we skip this entire page as well. To synchronize the row reading we have to skip the
+     * values (and the related rl and dl) for the rows [20, 39] in the end of the page 0 for col2. Similarly, we have to
+     * skip values while reading page0 and page1 for col3.
+     */
+    private void processValuesSync(int valuesToRead, Consumer<Void> valueConsumer)
+    {
+        if (definitionLevel == EMPTY_LEVEL_VALUE && repetitionLevel == EMPTY_LEVEL_VALUE) {
+            definitionLevel = definitionReader.readLevel();
+            repetitionLevel = repetitionReader.readLevel();
+        }
+        int valueCount = 0;
+        int skipCount = 0;
+        for (int i = 0; i < valuesToRead; ) {
+            boolean consumed = false;
+            do {
+                if (nextRow(repetitionLevel)) {
```

Comment:
I'm not sure I understand the semantics of the `nextRow` method. What does it mean for it to return `true` vs `false`? Why do we skip a value when it returns `true`?

Reply:
Added comment. We skip rows until the currentRow reaches the targetRow.

Reply:
Also, renamed function to make it more clear what it's doing.